### ISSUE TITLE
Configure the interrupt handler and getting the interrupt 

### ISSUE THREAD

**Issue by DevendraDevadiga** [OP-TEE/optee_os#4691]
Dear Sir,
I am trying to get the interrupt in OP-TEE in i.MX8MQ. But as per the suggestion provided in below link we need to configure the GIC in ATF (Arm-trusted-firmware) and register the interrupt handler in ATF.
https://community.nxp.com/t5/i-MX-Processors/Configure-Shared-Peripheral-Interrupt-as-Secure-to-Handle-in-S/m-p/986442#M146495

As per the suggestion I tried to configure the interrupt 64 as in Secure Group 0. In i.MX8MQ I am trying for GPIO1_6 and IRQ number is 64.
In the discussion mentioned like we need to add 32 for interrupt number i.e 64+32.

_static const interrupt_prop_t g01s_interrupt_props[] = {
INTR_PROP_DESC(6, GIC_HIGHEST_SEC_PRIORITY,
INTR_GROUP1S, GIC_INTR_CFG_LEVEL),
INTR_PROP_DESC(7, GIC_HIGHEST_SEC_PRIORITY,
INTR_GROUP0, GIC_INTR_CFG_LEVEL),
**INTR_PROP_DESC( (64 + 32), GIC_HIGHEST_SEC_PRIORITY, INTR_GROUP0, GIC_INTR_CFG_LEVEL), // <-- my entry here**
};_

Interrupt registration implementation added by refering below source code:

https://github.com/ARM-software/arm-trusted-firmware/blob/master/plat/nxp/common/setup/ls_interrupt_mgmt.c

_$cp "plat/nxp/common/setup/ls_interrupt_mgmt.c" "plat/imx/common/ls_interrupt_mgmt.c"
$cp "plat/nxp/common/setup/include/ls_interrupt_mgmt.h" "./plat/imx/common/include/ls_interrupt_mgmt.h"_

Just after the GICv3 driver initialization, implemented the test interrupt handler for IRQ 64 i.e (64+32).

_#ifdef IW_TEST_INTERRUPT
**static uint64_t int64_interrupt_handler(uint32_t id, uint32_t flags, void *handle, void *cookie)
{
ERROR("int64_interrupt_handler..!!!! id = %d \n", id);
return 0;
}**
#endif_

_void bl31_platform_setup(void)
{
generic_delay_timer_init();
/* init the GICv3 cpu and distributor interface */
plat_gic_driver_init();
plat_gic_init();
/* gpc init */
imx_gpc_init();
dram_info_init(SAVED_DRAM_TIMING_BASE);
#ifdef IW_TEST_INTERRUPT
ERROR(" ls_el3_interrupt_config\n");
**ls_el3_interrupt_config();
request_intr_type_el3((64 + 32), int64_interrupt_handler);**
#endif

}_

In OP-TEE GIC configuration every config will be overwrite. So I commented the GIC configuration in OP-TEE.

Once booted with above changes, just after Linux OP-TEE Driver initialization I am getting interrupt prints continuously as below:

__[ 2.902233] caam-snvs 30370000.caam-snvs: violation handlers armed - init state
[ 2.910650] hidraw: raw HID events driver (C) Jiri Kosina
[ 2.926934] usbcore: registered new interface driver usbhid
[ 2.932603] usbhid: USB HID core driver
[ 2.942859] usb 1-1: New USB device found, idVendor=0424, idProduct=2514
[ 2.949965] usb 1-1: New USB device strings: Mfr=0, Product=0, SerialNumber=0
[ 2.976422] ashmem: initialized
**INFO: Interrupt recvd is 96
ERROR: int64_interrupt_handler..!!!! id = 96
INFO: Interrupt recvd is 96
ERROR: int64_interrupt_handler..!!!! id = 96
INFO: Interrupt recvd is 96
ERROR: int64_interrupt_handler..!!!! id = 96
INFO: Interrupt recvd is 96**
..................._

What is the reason I am getting interrupt continuosly here ?

Interrupt configuration is done as below:

**_set_interrupt_rm_flag(flags, NON_SECURE);
rc = register_interrupt_type_handler(INTR_TYPE_EL3,
ls_el3_interrupt_handler, flags);_**

Below is the code of interrupt handler:

_**static uint64_t ls_el3_interrupt_handler(uint32_t id, uint32_t flags,
void *handle, void *cookie)
{
uint32_t intr_id;
interrupt_type_handler_t handler;
intr_id = plat_ic_get_pending_interrupt_id();
INFO("Interrupt recvd is %d\n", intr_id);
handler = type_el3_interrupt_table[intr_id];
if (handler != NULL) {
handler(intr_id, flags, handle, cookie);
}

plat_ic_end_of_interrupt(intr_id);
return 0U;
}**_

I had attached the log for reference. Please can you suggest me what changes required for getting interrupt properly ?

Thanks and Regards,
Devendra

[el3_interrupt.txt](https://github.com/OP-TEE/optee_os/files/6686375/el3_interrupt.txt)

**Comment by jenswi-linaro** [comment]
It looks like you've configured TF-A to recieve the interrupt, but you don't acknowledge it so it triggers again, and again.

**Comment by DevendraDevadiga** [comment]
I tried to add acknowledge function as below:


_static uint64_t ls_el3_interrupt_handler(uint32_t id, uint32_t flags,
                                          void *handle, void *cookie)
{
        uint32_t intr_id;
        uint32_t intr_raw;
        interrupt_type_handler_t handler;
        INFO(" ls_el3_interrupt_handler\n");
        //intr_id = plat_ic_get_pending_interrupt_id();
        /* Acknowledge IRQ */
        **intr_raw = plat_ic_acknowledge_interrupt();**
        INFO(" intr_raw: %d\n", intr_raw);
        **intr_id = plat_ic_get_interrupt_id(intr_raw);**
        INFO("Interrupt recvd is %d\n", intr_id);
        handler = type_el3_interrupt_table[intr_id];
        if (handler != NULL) {
                handler(intr_id, flags, handle, cookie);
        }
        /*
         * Mark this interrupt as complete to avoid a interrupt storm.
         */
        plat_ic_end_of_interrupt(intr_id);
        return 0U;
}_

This time I didn't get the inetrrupt continuosly as in previous. But plat_ic_get_interrupt_id() returns 0 instead 96.

_[    3.283060] mmcblk0boot0: mmc0:0001 DG4008 partition 1 4.00 MiB
**INFO:     ls_el3_interrupt_handler
INFO:     intr_raw: 0
INFO:    Interrupt recvd is 0**
[    3.299942] mmcblk0boot1: mmc0:0001 DG4008 partition 2 4.00 MiB
[    3.302923] optee: probing for conduit method from DT.
I/[    3.306369] mmcblk0rpmb: mmc0:0001 DG4008 partition 3 4.00 MiB
TC:  Dynamic shared memory is disabled
[    3.321672] optee: initialized driver_


Actually I used the GPIO lines which is connected to touch interrupt line. Whenevr I touch the device interrupt should be captured.

But here I am not getting any interrupt.

My intention is for specified gpio number I should be able to register the interrupt in OP-TEE OS and when level/edge trigger happens I should be able to capture the interrupt in OP-TEE OS.

I am using i.MX8MQ (Cortex A53) and enabled "OPTEED" in ATF.
For configuring GICV3 and register the iterrupt handler what is the proper method ?

In this link they mentioned like we need to configure the GIC in ATF and interrupt as Group 0 and we need to register the handler in ATF itself.  Is this proper ?
https://community.nxp.com/t5/i-MX-Processors/Configure-Shared-Peripheral-Interrupt-as-Secure-to-Handle-in-S/m-p/986442#M146495

Example I2C driver is in OP-TEE and if I register the handler in ATF, how we can forward the interrupt to OP-TEE?

Or do we configure the GICV3 in ATF and register the handler in OP-TEE?
Or do we configure the both GICV3 and register the handler in OP-TEE ?

Please can you suggest me what is the proper method ?

Can you provide the refernce to example source code where GIC is configured and interrupt handler is registered (Where ATF and OPTEE is used). In this case we need to disable/modify the GIC driver in Linux ? Because Linux GIC driver also doing some of GICV3 configuration. 

I know how to configure the GIC and enable the interrupt handler in CortexA7 case. In that gic is configured in OP-TEE and interrupt handler also such that we are able to capture the interrupt.

Regards,
Devendra Devadiga

**Comment by jenswi-linaro** [comment]
Here's an example for QEMU-v8, FVP and Juno (type in the secure console to trigger interrupts). It depends on GICv2 so as the comment suggest you need to update the dispatcher to know to forward specific interrupts to S-EL1.
https://github.com/OP-TEE/optee_os/blob/3ddd5cd700ed03e2265c9c3286697fd3c395a7bc/core/arch/arm/plat-vexpress/main.c#L92-L128

**Comment by DevendraDevadiga** [comment]
Hi,
Thanks you for your support.
Is this possible to provide the reference to an ATF source code correspond to above OP-TEES OS where interrupt (IT_CONSOLE_UART i.e interrupt number 38) is configured in ATF. ?

@you need to update the dispatcher to know to forward specific interrupts to S-EL1
Is this already available in OP-TEE dispatcher or specific to our interrupt number we need to do any implementation?
How this is done in case of QEMU-v8, FVP and Juno in ATF?
Regards,
Devendra

**Comment by DevendraDevadiga** [comment]
In "main_init_gic()" its mentioned as below:
#if defined(PLATFORM_FLAVOR_fvp) || defined(PLATFORM_FLAVOR_juno) || \
        defined(PLATFORM_FLAVOR_qemu_armv8a)
        ****/* On ARMv8, GIC configuration is initialized in ARM-TF */****
        gic_init_base_addr(&gic_data, gicc_base, gicd_base);
#else
        /* Initialize GIC */
        gic_init(&gic_data, gicc_base, gicd_base);
#endif

Can I know corresponding ARM-TF source code.

Thanks and Regards,
Devendra

**Comment by jenswi-linaro** [comment]
This is where stuff is registered with GICv2 in TF-A. What you need to do for GICv3 I'm afraid you need to figure out yourself.
https://github.com/ARM-software/arm-trusted-firmware/blob/967344b52082725523ed6146b54e8c8a077bfeb6/services/spd/opteed/opteed_main.c#L286-L307

**Comment by DevendraDevadiga** [comment]
Dear Sir,
I checked in "plat-synquacer" in OP-TEE OS source code. There ARM GICV3 is enabled.
https://github.com/OP-TEE/optee_os/blob/master/core/arch/arm/plat-synquacer/conf.mk#L22
And Interrupt is registered here for timer:
https://github.com/OP-TEE/optee_os/blob/master/core/arch/arm/plat-synquacer/main.c#L79
 Similary, in ATF they configured the interrupt for timer here:

https://github.com/ARM-software/arm-trusted-firmware/blob/master/plat/socionext/synquacer/sq_gicv3.c#L32

But I don't think they added any changes in OP-TEE Dispatcher related to GICV3 case. In OP-TEE dispatcher I think the implementation is there for handling FIQ/IRQ wrt to GICV3 ie IRQ as secure interrupt and FIQ as Non-secure interrupt.
Is my understanding correct ?

**Comment by jenswi-linaro** [comment]
It looks like you've found a working example with GICv3.

> But I don't think they added any changes in OP-TEE Dispatcher related to GICV3 case.

My bad.

> Please can you check this and explain me what they did to get the interrupt in GICV3 case?

You have a nice example which shouldn't be too hard to follow given what you're trying to do.

**Comment by DevendraDevadiga** [comment]
Dear Sir,

Thank for your reply.
I need some clarification related to OP-TEE version, Linux Kernel version and OP-TEE Linux Driver.
Currently I am using Android Pie with Linux kernel version as 4.14.98 and OP-TEE version is 3.2.

- Can we use OP-TEE 3.13 along with Linux 4.14.98? 
- In this case we need to update OP-TEE Linux Driver also? If yes,can provde reference to OP-TEE Linux driver with respect to OP-TEE 3.13 and Linux 4.14.98 ?

Regards,
Devendra

**Comment by jenswi-linaro** [comment]
> * Can we use OP-TEE 3.13 along with Linux 4.14.98?
> * In this case we need to update OP-TEE Linux Driver also? If yes,can provide reference to OP-TEE Linux driver with respect to OP-TEE 3.13 and Linux 4.14.98 ?

It sounds like it should work.
